velox/functions/sparksql/DateTimeFunctions.cpp

```diff
@@ -0,0 +1,161 @@
+      seconds);
+  if (seconds == 60) {
+    VELOX_USER_CHECK(
+        nanos == 0,
+        "The fraction of sec must be zero. Valid range is [0, 60].");
+  }
+
+  auto daysSinceEpoch = util::daysSinceEpochFromDate(
+      year->valueAt<int32_t>(row),
+      month->valueAt<int32_t>(row),
+      day->valueAt<int32_t>(row));
+  auto localMicros = hour->valueAt<int32_t>(row) * util::kMicrosPerHour +
+      minute->valueAt<int32_t>(row) * util::kMicrosPerMinute +
+      micros->valueAt<int64_t>(row);
+  return util::fromDatetime(daysSinceEpoch, localMicros);
+}
+
+class MakeTimestampFunction : public exec::VectorFunction {
+ public:
+  MakeTimestampFunction() = default;
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    auto microsType = args[5]->type()->asShortDecimal();
+    VELOX_USER_CHECK(
+        microsType.scale() == 6,
+        "Seconds fraction must have 6 digits for microseconds but got {}",
+        microsType.scale());
+
+    context.ensureWritable(rows, TIMESTAMP(), result);
+    auto* resultFlatVector = result->as<FlatVector<Timestamp>>();
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto year = decodedArgs.at(0);
+    auto month = decodedArgs.at(1);
+    auto day = decodedArgs.at(2);
+    auto hour = decodedArgs.at(3);
+    auto minute = decodedArgs.at(4);
+    auto micros = decodedArgs.at(5);
+
+    if (args.size() == 7) {
+      // If the timezone argument is specified, treat the input timestamp as the
+      // time in that timezone.
+      if (args[6]->isConstantEncoding()) {
+        auto constantTzID =
+            util::getTimeZoneID(args[6]
+                                    ->asUnchecked<ConstantVector<StringView>>()
+                                    ->valueAt(0)
+                                    .str());
+        rows.applyToSelected([&](vector_size_t row) {
+          auto timestamp = makeTimeStampFromDecodedArgs(
+              row, year, month, day, hour, minute, micros);
+          timestamp.toGMT(constantTzID);
+          resultFlatVector->set(row, timestamp);
+        });
+      } else {
+        auto timeZone = decodedArgs.at(6);
+        rows.applyToSelected([&](vector_size_t row) {
+          auto timestamp = makeTimeStampFromDecodedArgs(
+              row, year, month, day, hour, minute, micros);
+          auto tzID =
+              util::getTimeZoneID(timeZone->valueAt<StringView>(row).str());
+          timestamp.toGMT(tzID);
+          resultFlatVector->set(row, timestamp);
+        });
+      }
+    } else {
+      // Otherwise use session timezone. If session timezone is not specified,
+      // use default value 0(UTC timezone).
+      int64_t sessionTzID = 0;
+      const auto& queryConfig = context.execCtx()->queryCtx()->queryConfig();
+      const auto sessionTzName = queryConfig.sessionTimezone();
```

Comment:
By declaring this function as stateful, sessionTzName could become a construction parameter of MakeTimestampFunction, instead of being calculated for each apply.